train.py

```diff
@@ -735,6 +795,55 @@ def train_one_epoch(
         if model_ema is not None:
             model_ema.update(model)
 
+        if loss_ is not None:
+            loss_.copy_(loss)
+            return None
+        else:
+            return loss
+
+    for batch_idx, (input, target) in enumerate(loader):
+        last_batch = batch_idx == last_idx
+        data_time_m.update(time.time() - end)
+        if not args.prefetcher:
+            input, target = input.cuda(), target.cuda()
+            if mixup_fn is not None:
+                input, target = mixup_fn(input, target)
+
+        if cg_stage is None:    # The default non-CUDAGraph mode
+            loss = _step(input, target)
+        elif cg_stage == 'get_static_shapes':
+            s = torch.cuda.Stream()
+            s.wait_stream(torch.cuda.current_stream())
+            with torch.cuda.stream(s):
+                optimizer.zero_grad(set_to_none=True)
+                loss = _step(input, target)
+            torch.cuda.current_stream().wait_stream(s)
+            return (input, target, loss)
+        elif cg_stage == 'capture':
+            s = torch.cuda.Stream()
+            s.wait_stream(torch.cuda.current_stream())
+            with torch.cuda.stream(s):
+                for _ in range(11):
+                    optimizer.zero_grad(set_to_none=True)
+                    _step(input, target)
+            torch.cuda.current_stream().wait_stream(s)
+
+            g = torch.cuda.CUDAGraph()
+            optimizer.zero_grad(set_to_none=True)
+            with torch.cuda.graph(g):
+                _step(cg_static_input, cg_static_target, cg_static_loss, scaler_step_and_update=False)
+            return g
+        elif cg_stage == 'replay':
+            cg_static_input.copy_(input)
+            cg_static_target.copy_(target)
+            cuda_graph.replay()
+            loss = cg_static_loss
+            if loss_scaler is not None:
+                loss_scaler.step_and_update()
+
+        if not args.distributed:
+            losses_m.update(loss.item(), input.size(0))
+
         torch.cuda.synchronize()
```

Comment:
In order to amortize CPU overhead by allowing it to run ahead during the previous step, I wanted to suggest not syncing on every step and instead recording the time at the beginning and end of the run and dividing by the batch size.

Reply:
Agreed, this is typically a better approach and more representative of the user experience.